### ISSUE TITLE
FEC generator is not working

### ISSUE THREAD

**Issue by herrfrei** [ibreakifix/PorschePCMStuff#2]
ibreakifix, I found some problems with the generator and also prepared a Python FEC parser. Should I create a pull request?
And, I'd like to discuss some things with you. Do you have an email address (or some other private contact option) for me?

**Comment by ibreakifix** [comment]
Yeah just do a pull request and I can add it in. Once you generate the new fec-container.fec file, you'll need to patch out the signature check in MIBRoot, which I can't offer any guidance on at the moment.

**Comment by askac** [comment]
It works as I tested.
But need some patch in the generating script.

**Comment by askac** [comment]
Sorry
Today I test again and found it not work :(
I think maybe something is checking the key internally.

And, change the key will cause your navigation malfunction because the navigation data were also signed. This would be a big task to replace the key...

To have "carPlay", the best way is to patch the smartphone_ixxx (I don't remember name) in mmx's app directory... For other functions, patch MIBRoot seems a better way. (What you need is patch the signature check....)

**Comment by ibreakifix** [comment]
Right, the key gets checked against another key stored in the OTP area of your NOR flash, so the original attempt of replacing the key files won't work.

So after creating your new FEC file, your old keys remain but MIBRoot will need to be patched to accept the improperly signed FEC file. I'm not the right person to instruct on how to do that, so hopefully somebody else will share.

**Comment by askac** [comment]
There is a key word - checkSecMSignature. I think the FEC check is performed here.
But I think use exception list is better. Key work - checkSHA1Signature.
You can refer to the prodEL to write the exception list. Just need to patch the checkSHA1Signature.
Leave original FEC there.

**Comment by herrfrei** [comment]
Good idea, askac! The prodEL.txt file should be located in /mnt/efs-persist/etc/prodEL.txt, right? And, do you know how to patch the MIBroot?

**Comment by navidoctor** [comment]
Hi, I've been working on  these systems for quite some time. 

The keys are not stored in the OTP, they are inside the NEC CPU. At least the metainfo key is. If the version of flash for the required Firmware does not match the NEC it will show metakey error, so this key is checked by the NEC cpu against the keys stored in the unit. As soon as you replace the flash from the NEC with the right one that belongs to the firmware your "KEY" problem is solved. 


The solution should be found there if you want to use it without hacking the MIBROOT.

Regards

**Comment by askac** [comment]
There would be a perfect method if you can create a sha1 checksum collision on ExceptionList.txt

To get the sha1sum of ExceptionList. You can use command
"cat ExceptionList.txt|grep -iv signature|sha1sum"
In my unit, the sha1 checksum is 
d4d6eda34beb3ad1ec8acbbeaeca42f5d05ac62c

Same as decoded from signature and public key.
The expected format could be found in prodEL.txt.
Example.
# Exception list for development purpose (MIB2-High)

#StartOfFazitIDs
???-?????.??.?????????? #Match any FazitID
#EndOfFazitIDs

# Boardbook data path
Destination = "/net/mmx/mnt/boardbook"

# PPOI data path
Destination = "/net/mmx/mnt/navdb/PersonalPOI"

# List of supported FSCs

[SupportedFSC]
fsc1 = "00030000"
fsc2 = "00040100"
fsc3 = "00050000"
fsc4 = "00060100"
fsc5 = "00060200"
fsc6 = "00060300"
fsc7 = "00060400"
fsc8 = "00060500"
fsc9 = "00060600"
fsc10 = "00060700"
fsc11 = "00060800"
fsc12 = "00060900"
fsc13 = "00060A00"
fsc14 = "00070100"
fsc15 = "00070200"
fsc16 = "00070400"
fsc17 = "02300016"
fsc18 = "02310016"
fsc19 = "02320022"
fsc20 = "02330016"
fsc21 = "023A0016"
fsc22 = "023B0016"
fsc23 = "023D0016"
fsc24 = "03300016"
fsc25 = "03310016"
fsc26 = "03320016"
fsc27 = "03330016"
fsc28 = "033A0016"
fsc29 = "033B0016"
fsc30 = "033D0016"
fsc31 = "06100004"
fsc32 = "06300016"
fsc33 = "06310016"
fsc34 = "06320016"
fsc35 = "06330016"
fsc36 = "063A0016"
fsc37 = "063B0016"
fsc38 = "063D0016"
fsc39 = "07300016"
fsc40 = "073D0016"
fsc41 = "08300003"
fsc42 = "083D0003"
fsc43 = "08320001"
fsc44 = "09300008"
fsc45 = "09310002"
fsc46 = "09320008"
fsc47 = "09330003"
fsc48 = "093A0000"
fsc49 = "093B0004"
fsc50 = "093D0002"
...(List all you want. If you are kind, you can list all brand and please share)

#HERE THE GARBAGE BLOCK TO GENERATE sha1 collision
#For this signature is "d4d6eda34beb3ad1ec8acbbeaeca42f5d05ac62c"
#Then append following Signature. Check with "cat ExceptionList.txt|grep -iv signature|sha1sum"
[Signature]
signature1 = "8b7578817a3f0ae512f4d3e7cc52c0c2"
signature2 = "d86089dec32d22f6f6d4cd2a6d1fafed"
signature3 = "7e6bb811302deafef2974a5956dc74a1"
signature4 = "75c9a52f8ff5211f27f068b6edd63318"
signature5 = "2285de6bdbc51e4a22c17d43215a8b67"
signature6 = "dcdcadd6f6cbaa3c4b6c3c3e592e7b1b"
signature7 = "a8fb501242f2bb2a3cc48ff0659c48c7"
signature8 = "b1eeb65c7e01aea0e65e00c32b63518d"

**Comment by askac** [comment]
Extra information: How to get the sha1 checksum from signature
Use Python:
hex(pow(0x8b7578817a3f0ae512f4d3e7cc52c0c2d86089dec32d22f6f6d4cd2a6d1fafed7e6bb811302deafef2974a5956dc74a175c9a52f8ff5211f27f068b6edd633182285de6bdbc51e4a22c17d43215a8b67dcdcadd6f6cbaa3c4b6c3c3e592e7b1ba8fb501242f2bb2a3cc48ff0659c48c7b1eeb65c7e01aea0e65e00c32b63518d,3)%0xc0f389eec7b66c9dc736508ff88aeb1fb113942ead020814d08d29e868f14b2086bcd7ddccba7559f999e76d24619660bbe17434da59988087f2a99cd465b1ff423522b78cb0de463a669613d356dfa9e86e0e2e0b6dab5de89131c5a0727aeab1767278ab101dcd9c3cfc1026705c1dab3bf53bf50afafb3f52da2ceb0bee57)
Get result:
'0x1ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff003021300906052b0e03021a05000414d4d6eda34beb3ad1ec8acbbeaeca42f5d05ac62cL'

echo 3021300906052b0e03021a05000414d4d6eda34beb3ad1ec8acbbeaeca42f5d05ac62c|xxd -r -p |openssl asn1parse -inform DER -dump

The result
    0:d=0  hl=2 l=  33 cons: SEQUENCE
    2:d=1  hl=2 l=   9 cons: SEQUENCE
    4:d=2  hl=2 l=   5 prim: OBJECT            :sha1
   11:d=2  hl=2 l=   0 prim: NULL
   13:d=1  hl=2 l=  20 prim: OCTET STRING
      0000 - d4 d6 ed a3 4b eb 3a d1-ec 8a cb be ae ca 42 f5   ....K.:.......B.
      0010 - d0 5a c6 2c                                       .Z.,

**Comment by askac** [comment]
I think other signature contains other sha1sum should be also work if you can find a collision with sha1 checksum contained in other signature (often you can get from metainfo2.txt of map installation data medium).

**Comment by ibreakifix** [comment]
@askac How feasible is a collision on the SHA1 checksum? I'm really doubtful that we can calculate the required garbage bytes with a reasonable amount of computing resources. I think a more pracitcal approach is to change the FAZIT ID and recalculate the checksum, then upload a valid prodEL.txt from a preproduction unit. For example (signature appears to be invalid though - perhaps signed w/ MIB1 MI Key?):
`# Exception list for HAD production mode tests

#StartOfFazitIDs
JTB-?????.??.11????????    # MIB1 universal FAZIT string
BKA-?????.??.12????????    # MIB2 universal FAZIT string
#EndOfFazitIDs

# Boardbook data path
Destination = "/net/mmx/mnt/boardbook"

# PPOI data path
Destination = "/net/mmx/mnt/navdb/PersonalPOI"


# List of supported FSCs

[SupportedFSC]
fsc   = "00030000"                  # AMI
fsc2  = "00030001"                  # Gracenote
fsc3  = "00040100"                  # Navigation
fsc4  = "00050000"                  # Bluetooth
fsc5  = "00060100"
fsc6  = "00060200"
fsc7  = "00060300"
fsc8  = "00060400"
fsc9  = "00060500"
fsc10 = "00060600"
fsc11 = "00060700"
fsc12 = "00060800"
fsc13 = "00060900"
fsc14 = "00070100"                  # SDS
fsc15 = "00070200"                  # SDS for Nav

# Audi MIB1 NavDB
fsc16 = "02100000;0210003F"         # EU
fsc17 = "02110000;0211003F"         # NAR
fsc18 = "021D0000;021D003F"         # ROW

# Audi MIB2 NavDB
fsc19 = "02300000;0230003F"         # EU
fsc20 = "02310000;0231003F"         # NAR
fsc21 = "023D0000;023D003F"         # ROW

# Bentley MIB2 NavDB
fsc22 = "03300000;0330003F"         # EU
fsc23 = "03310000;0331003F"         # NAR
fsc24 = "033D0000;033D003F"         # ROW

# Porsche MIB1 NavDB
fsc25 = "06100000;0610003F"
fsc26 = "06110000;0611003F"
fsc27 = "061D0000;061D003F"

# Porsche MIB2 NavDB
fsc28 = "06300000;0630003F"         # EU
fsc29 = "06310000;0631003F"         # NAR
fsc30 = "063D0000;063D003F"         # ROW

# Seat MIB1 NavDB
fsc31 = "07100000;0710003F"

# Seat MIB2 NavDB
fsc32 = "07300000;0730003F"         # EU
fsc33 = "073D0000;073D003F"         # ROW

# Skoda MIB1 NavDB
fsc34 = "08100000;0810003F"         # EU
fsc35 = "08110000;0811003F"         # NAR
fsc36 = "08140000;0814003F"         # AGCC
fsc37 = "08140100;0814013F"         # AGCC, Turkey
fsc38 = "08140200;0814023F"         # AGCC, Israel
fsc39 = "08140300;0814033F"
fsc40 = "08150000;0815003F"         # South Africa
fsc41 = "08160000;0816003F"         # Australia, New Zealand
fsc42 = "08170000;0817003F"         # India
fsc43 = "08180000;0818003F"         # Mexico
fsc44 = "08180100;0818013F"         # Chile
fsc45 = "08190000;0819003F"         # Asia/Pacific
fsc46 = "081D0000;081D003F"         # ROW
fsc47 = "081D0100;081D013F"

# Skoda MIB2 NavDB
fsc48 = "08300000;0830003F"         # EU
fsc49 = "083D0000;083D003F"         # ROW

# VW MIB1 NavDB
fsc50 = "09100000;0910003F"         # EU
fsc51 = "09110000;0911003F"         # NAR
fsc52 = "09140000;0914003F"         # Turkey
fsc53 = "09140100;0914013F"         # AGCC, Turkey
fsc54 = "09140200;0914023F"         # AGCC, Israel
fsc55 = "09140300;0914033F"         # AGCC
fsc56 = "09150000;0915003F"         # South Africa
fsc57 = "09160000;0916003F"         # Australia, New Zealand
fsc58 = "09170000;0917003F"         # India
fsc59 = "09180000;0918003F"         # Argentina, Brazil, Mexico
fsc60 = "09180100;0918013F"         # Chile
fsc61 = "09190000;0919003F"         # Asia/Pacific
fsc62 = "091D0000;091D003F"         # ROW

# VW MIB2 NavDB
fsc63 = "09300000;0930003F"         # EU
fsc64 = "09310000;0931003F"         # NAR
fsc65 = "093D0000;093D003F"         # ROW

#EndOfFSCs


[Signature]
signature1 = "5aa973fa008eedf98f75471c166de95a"
signature2 = "76592bfc6db54949f557c2261390a4ad"
signature3 = "42413e9f5f3fb7ead0d194c58b518d85"
signature4 = "615ab87f313b8e90a27d7d4926eca547"
signature5 = "8afc3f6c7d1e9d6799ae9d71285be56d"
signature6 = "0f2c4babb8f2cf9176e60b84e5adb36c"
signature7 = "81e8301d8ffcd527276660e04ca51c27"
signature8 = "557143257b7e07ac39fe3729da04472f"
`

**Comment by jilleb** [comment]
[SupportedFSC]
fsc = "00030000" # AMI
fsc2 = "00030001" # Gracenote
fsc3 = "00040100" # Navigation
fsc4 = "00050000" # Bluetooth
fsc5 = "00060100"# Fahrzeugdatenschnittstelle (SAI_VehicleData)
fsc6 = "00060200" # Infotainmentsteuerung (SAI_RSE), rear seat entertainment
fsc7 = "00060300" #Mirror Link
fsc8 = "00060400" #Sport-HMI
fsc9 = "00060500" #Sport Chrono
fsc10 = "00060600" #Fahrtenbuch
fsc11 = "00060700" #Mobile Online Dienste (MOD) im HMI
fsc12 = "00060800" #Apple Carplay
fsc13 = "00060900" #Google Automotive Link / Android Auto
fsc14 = "00070100" # SDS
fsc15 = "00070200" # SDS for Nav

**Comment by navidoctor** [comment]
Hi, good evening :)

Where is this VRCN found ? I know its always in the FSC but where is it found or how is it calculated ?

Regards

**Comment by askac** [comment]
VCRN could be found from original fec or by measurement function of ODIS or other tool.
Named "Individualization characteristic" (or words like this)

**Comment by askac** [comment]
@ibreakifix
As I know there is some research on sha1 collision.
Modifying FazitID could cause a lot of problem. I think maybe only patch MIBRoot is enough.
The keywork is "checkSHA1Signature"
You can search the string if you have  IDA pro.

There is a statement **v4 = memcmp(....) == 0;**
Make a patch on the statement to make it become **v4 = 1** .

Then you can use the dumpifs -u to decompress the flash image.
Modify decompressed image.
Repack it, flashit. (Tools available on my github repository)

After patched image had flashed.
Modify the ExceptionList. Put back. 
Everything should work fine.

**Comment by askac** [comment]
@ibreakifix
And I also checked the  signature of prodEL.txt
The signature was not signed by the internal data key.
I think change FazitID may not work.

**Comment by maxiMIB2** [comment]
In my opinion is ExceptionList + pached image the easiest way (you don't need change FaziID or something)... I use it on my unit for quite long time now without any trouble. If I have a little more time I'll take a look at the FSC generator and dumper because currently it's not running with my FSC.

**Comment by navidoctor** [comment]
HI, 

I tried to use the script with the given commands. I installed openssl in ubuntu, somehow I get errors : sh MIB2_FEC_Generator.sh -f 00040000 -n 12345678 -v WAUZZZFYXJ1234567 -d /media/IC7A-D5BC/FEC

Error are these :

: cannot execute - No such file or directory
: cannot execute - No such file or directory
: cannot execute - No such file or directory
: cannot execute - No such file or directory
: cannot execute - No such file or directory
: cannot execute - No such file or directory
: cannot execute - No such file or directory
: cannot execute - No such file or directory
: cannot execute - No such file or directory
: cannot execute - No such file or directory

Any idea what I am doing wrong ?

Regards

**Comment by ibreakifix** [comment]
Does the folder "/media/IC7A-D5BC/" exist? Try to create the FEC files on your home directory "~" and then move them to your target folder

**Comment by navidoctor** [comment]
Hi, 

Yes, i did both, checked the folder, checked permissions.

Any other programs that should be installed that i am missing ?

regards

**Comment by ibreakifix** [comment]
Working fine on Ubuntu 19.04 for me. I can't see where the script is failing in your errors, so I can't make a fix. You could try to redownload the repo as a zip, then run the script from the extracted file.... it's possible something is messed up if you created the .sh file from copy and pasting.

**Comment by navidoctor** [comment]
HI

sh MIB2_FEC_Generator.sh -f 00040100 -n 12345678 -v WAUZZZFY4Z1234567

MIB2_FEC_Generator.sh: 125: MIB2_FEC_Generator.sh: Syntax error: redirection unexpected

This is what happens also with -d option when i direct to valid folder

Script is downloaded and taken out of fresh winrar.

Regards

**Comment by ibreakifix** [comment]
Run the script from bash and not sh

**Comment by navidoctor** [comment]
Thanks, this makes the script run. Sorry for the mistake.

Only problem now is that all keys etc are made except the actual FEC container is besides 8 bytes empty.

The whole script runs nice only problem is that result is nearly empty FEC

Ant idea ?

**Comment by pindour** [comment]
Hello,
please contact us about cooperation at bawo at centrum.cz

**Comment by aigortt** [comment]
> Hola,
> ibreakifix ... askac .... por favor contáctenos sobre cooperación en bawo en centrum.cz

¿quieres que hablemos?
